### PR TITLE
New quickstart: Errai Hello World

### PR DESCRIPTION
**Pull request by jfuerth**
This is the replacement for pull request #99, but this time with a feature branch so we can update with changes. (oops, I'm a pull request n00b)

This request includes fixes for some of the issues identified by Pete Muir in the original pull request.

Outstanding issues not yet addressed:
- create a GWT+Errai BOM and inherit the GWT and Errai version numbers from that
- release Errai 2.0.Beta1 and refer to that version of Errai from this project's pom
- publish Errai 2.0.Beta1 to Maven Central and remove reference to JBoss Community repo

### CODE REVIEW

**Comment by sgilda** [comment]
Hi Jonathon,
I'm not great at git, so maybe I'm doing something stupid, but when I look at your repository here: https://github.com/jfuerth/quickstart, I see an errai_kitchensink branch, but I don't see a errai_helloworld branch. Any idea what I'm doing wrong?
Thanks, Sande

**Comment by sgilda** [comment]
I think I figured it out. It does appear to be the kitchensink-errai. Never mind. :-)
I'll look at it now.

**Comment by sgilda** [comment]
No luck.
    $ git remote add -f jfuerth https://github.com/jfuerth/quickstart.git
Updating jfuerth
remote: Counting objects: 90, done.
remote: Compressing objects: 100% (59/59), done.
remote: Total 88 (delta 11), reused 84 (delta 7)
Unpacking objects: 100% (88/88), done.
From https://github.com/jfuerth/quickstart
- [new branch]      errai_helloworld -> jfuerth/errai_helloworld
- [new branch]      errai_kitchensink -> jfuerth/errai_kitchensink
- [new branch]      master     -> jfuerth/master
  $ git checkout helloworld-errai jfuerth/errai_helloworld 
  error: pathspec 'helloworld-errai' did not match any file(s) known to git.
  error: pathspec 'jfuerth/errai_helloworld' did not match any file(s) known to git.
   $ git checkout kitchensink-errai jfuerth/errai_kitchensink  
  error: pathspec 'kitchensink-errai' did not match any file(s) known to git.
  error: pathspec 'jfuerth/errai_kitchensink' did not match any file(s) known to git.

Any idea what I'm doing wrong?

**Comment by jfuerth** [comment]
Hi Sande,

Probably the easiest way to get to my changes from where you are is the following:

git checkout -t jfuerth/errai_kitchensink

This will set up a "remote tracking" branch so you can pull or reset --hard to my changes as I respond to your code review with additional commits.

**Comment by jfuerth** [comment]
It's also useful to put this in your ~/.bash_profile so that you can see which branch you're on without typing "git status" as every second command. :)

function parse_git_branch {
  git branch --no-color 2> /dev/null | sed -e '/^[^_]/d' -e 's/_ (.*)/(\1)/'
}
PS1="\u@\h:\w\$(parse_git_branch)\$ "

**Comment by sgilda** [comment]
I just noticed that I missed the "-b" flag. Sorry about that! Thanks!

**Comment by sgilda** [comment]
Actually, I have this:
export PS1='\W$(__git_ps1 " (%s)") \$ '

I deleted that part of the display from the paste.
Should I change what I have to this?

**Comment by jfuerth** [comment]
It's "-t" that you need in this case. "git checkout -b" creates a new branch from the current commit, which is not what you want in this case.

If you've already done git checkout -b jfuerth/errai_kitchensink, you'll need to erase the new branch because its name will confuse git when you try to create the remote tracking branch. To erase that branch, do this:

$ git checkout master
$ git branch -d jfuerth/errai_kitchensink
$ git checkout -t jfuerth/errai_kitchensink

Then you should be in business. Isn't git fun? :)

Regarding your PS1 setting, if __git_ps1 resolves to the current branch name, you're all set. :)

**Comment by jfuerth** [comment]
Okay, one more comment here. In all of the above, I should have written errai_helloworld. I'm currently working on the kitchensink quickstart, but this pull request is for helloworld. :)

We'll be sending the kitchensink quickstart in a new pull request later this week.

**Comment by sgilda** [comment]
This tracking thing is new to me. I've been creating the branch, 
testing, pushing to my own repo to verify. I'll try this. Thanks!

**Comment by sgilda** [comment]
Got it!

**Comment by jfuerth** [comment]
Yay! \o/

**Comment by sgilda** [comment]
Hooray. I've finally got it! 

However, I get the following build error when I build it using mvn clean package.
[ERROR] Failed to execute goal on project jboss-as-helloworld-errai: Could not resolve dependencies for project org.jboss.as.quickstarts:jboss-as-helloworld-errai:war:7.0.2-SNAPSHOT: Failed to collect dependencies for [javax.enterprise:cdi-api:jar:1.0-SP4 (provided), org.jboss.spec.javax.annotation:jboss-annotations-api_1.1_spec:jar:1.0.0.Final (provided), com.google.gwt:gwt-user:jar:2.4.0 (provided), org.jboss.errai:errai-common:jar:2.0-SNAPSHOT (compile), org.jboss.errai:errai-bus:jar:2.0-SNAPSHOT (compile), org.jboss.errai:errai-ioc:jar:2.0-SNAPSHOT (compile), org.jboss.errai:errai-tools:jar:2.0-SNAPSHOT (compile), org.jboss.errai:errai-jaxrs-client:jar:2.0-SNAPSHOT (provided), org.jboss.errai:errai-jaxrs-provider:jar:2.0-SNAPSHOT (compile), org.jboss.resteasy:resteasy-jaxrs:jar:2.2.3.GA (provided)]: Failed to read artifact descriptor for org.javassist:javassist:jar:3.15.0-GA: Could not find artifact org.jboss:jboss-parent:pom:6-redhat-1 in jboss-public-repository-group (https://repository.jboss.org/nexus/content/groups/public/) -> [Help 1]
Did I miss something in the setup?

Also, if you don't mind, could you modify the readme title to conform to the format specified here https://docs.jboss.org/author/display/AS71/Writing+a+quickstart under "Each quickstart should contain a README.md (note that it should be named in UPPERCASE) markdown file that provides the following information about the quickstart "? Basically something like "helloworld-errai: Helloworld Using the Errai Framework"

Thanks again for the help!

**Comment by jfuerth** [comment]
Hi Sande,

I've updated the title in README.md to match the requirements.

As for the dependency problems, this is perplexing. I will try a build with an empty Maven cache and get back to you on this.

**Comment by sgilda** [comment]
Hi Jonathon,
There is only one errai quickstart on this page: https://docs.jboss.org/author/display/AS71/Contributing+a+Quickstart
It's number 52 and says it's errai-kitchensink (I'm assuming it should be kitchensink-errai to keep with the standard naming).
I don't see a helloworld-errai on that page. Should it be added?
Thanks, Sande

**Comment by jfuerth** [comment]
Yes, we only committed to the kitchensink quickstart, but based on emails I saw last week, it sounded like "the more the better" for the AS 7 quickstart collection. So we decided to submit both.

We also de have kitchensink-errai prepared. Based on Pete's comments regarding SNAPSHOT dependencies and reliance on the JBoss Community maven repo, we have held off on that pull request until the release of Errai 2.0 beta to Maven Central. We got Errai 2.0.Beta1 into the JBoss repo yesterday afternoon, and now we're working on Maven Central.

**Comment by sgilda** [comment]
This threw me a bit because the README.html file wasn't regenerated, but then I checked the README.md. :-)
Pete, I'm good with this one if you are.

**Comment by sgilda** [comment]
Pete, I generated the README.html file and pushed this to my github here: https://github.com/sgilda/quickstart/tree/helloworld-errai
If it looks OK to you, I can push to jbossas.

**Comment by sgilda** [comment]
Pete, do you need to review this one before we push it?

**Comment by jfuerth** [comment]
Hi Sande,

I still have to create the Errai/GWT BOM and update the POM in this pull request to reference it. I'll ping you back when that's done.

**Comment by sgilda** [comment]
OK. Thanks!

**Comment by sgilda** [comment]
I just tested this and it builds and deploys fine.

I'm a stupid user, so I don't know what to do with the GWT Development Mode window that pops up when I type: mvn gwt:run though. I assume anyone interested in this application will know what to do with it?

Also, I think some of the versions in the pom.xml file need to be updated to match https://docs.jboss.org/author/display/AS71/Maven+POM+Versions+Checklist. For example, org.jboss.as.quickstarts should be changed from 7.0.2-SNAPSHOT to 7.1.1-SNAPSHOT. A few others may need changes too.

@pmuir, does anyone need to do a code review for this one?

File: helloworld-errai/README.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+<p>You can also start JBoss AS 7 and deploy the project using Eclipse. See the JBoss AS 7
+Getting Started Guide for Developers for more information.</p>
+<h2>Running the Application in GWT Dev Mode</h2>
+<p>GWT Dev Mode provides an edit-save-refresh development experience. If you plan to try 
+modifying this demo, we recommend you start the application in Dev Mode so you don't 
+have to repackage the whole application every time you change it.</p>
+<p>Deploy the war file and start JBoss AS 7 as described above.</p>
+<p>Then execute the command:</p>
+<pre><code>mvn gwt:run
+</code></pre>
```
Comment (pmuir) [inline helloworld-errai/README.html]: We aren't committing the .html versions of files any more, could you remove this?

File: helloworld-errai/pom.xml [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      resources, i.e. build is platform dependent! -->
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+    <dependencyManagement>
+      <dependencies>
+        <dependency>
+          <groupId>org.jboss.bom</groupId>
+          <artifactId>jboss-javaee-6.0-with-errai</artifactId>
+          <version>1.0.0-SNAPSHOT</version>
```
Comment (pmuir) [inline helloworld-errai/pom.xml]: I've released 1.0.0.M4 of the BOMs which includes the errai BOM :-)

File: helloworld-errai/pom.xml [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      <url>https://repository.jboss.org/nexus/content/groups/public/</url>
+      <layout>default</layout>
+      <releases>
+        <updatePolicy>never</updatePolicy>
+      </releases>
+      <snapshots>
+        <updatePolicy>never</updatePolicy>
+      </snapshots>
+    </pluginRepository>
+  </pluginRepositories>
```
Comment (pmuir) [inline helloworld-errai/pom.xml]: Can this be removed now, as Errai is in central...

**Comment by pmuir** [comment]
Also, I still these files left on my disk after running mvn clean:

```
src/main/gwt-unitCache/
src/main/webapp/WEB-INF/deploy/
```

Can clean remove those as well?

**Comment by pmuir** [comment]
Jonathan, just a reminder that we're waiting for changes from you on this.

File: helloworld-errai/pom.xml (hunk 2 above)
Comment (jfuerth) [inline helloworld-errai/pom.xml]: The only problem is, Errai is now on 2.0.Beta3. The version in the BOM is 2.0.Beta2, and there's a critical bug fix in Beta3. We will send another BOM pull request.

Comment (pmuir) [inline helloworld-errai/pom.xml]: Np, I can release BOMs in ten mins so it's easy :-D

File: helloworld-errai/README.html (hunk 1 above)
Comment (jfuerth) [inline helloworld-errai/README.html]: removed.

File: helloworld-errai/pom.xml (hunk 2 above)
Comment (jfuerth) [inline helloworld-errai/pom.xml]: Pull request is here: https://github.com/jboss/jboss-bom/pull/6

Once this is released, I'll update this pull request.

File: helloworld-errai/pom.xml (hunk 3 above)
Comment (jfuerth) [inline helloworld-errai/pom.xml]: I created a new account on a Fedora box and I'm running a test build without the JBoss repos now.

**Comment by jfuerth** [comment]
Thanks again for the review, Pete. I think everything will be ready once we get the BOM up to Errai Beta 3.

**Comment by pmuir** [comment]
Merged. The sync to central won't happen til later but we can manage for a few hours :-)

**Comment by jfuerth** [comment]
Okay, I will reopen this pull request (to update to BOM M5) once the BOM update becomes visible in Central and I've had a good compile against it in a clean environment.

**Pull request closed without merging**